mlflow/types/utils.py

```diff
@@ -87,23 +87,28 @@ def _get_str_or_byte_type(data):
         return DataType.binary
 
 
-def _infer_datatype(data) -> Optional[DataType]:
+def _infer_datatype(data) -> DataType:
     if DataType.is_boolean(data):
         return DataType.boolean
-    if DataType.is_integer(data):
-        return DataType.integer
+    # Order of is_long & is_integer matters
+    # as both of theirs python_type is int
     if DataType.is_long(data):
         return DataType.long
-    if DataType.is_float(data):
-        return DataType.float
+    if DataType.is_integer(data):
+        return DataType.integer
+    # Order of is_double & is_float matters
+    # as both of theirs python_type is float
     if DataType.is_double(data):
         return DataType.double
+    if DataType.is_float(data):
+        return DataType.float
```

Comment:
what would happen if swap `is_double` and `is_float`?

Reply:
Then both `[1.0, 2.0]` and `[np.float32(0.1), np.float32(0.2)]` would be inferred as DataType.float, as the builtin python types are all float. But we should expect `[1.0, 2.0]` to be double, and `[np.float32(0.1), np.float32(0.2)]` to be float.